theano/gof/destroyhandler.py

```diff
@@ -947,10 +946,14 @@ def validate(self, fgraph):
                 if self.fail_validate:
                     err = self.fail_validate
                     self.fail_validate = False
-                    raise err
-            ords = self.orderings(fgraph)
-            if _contains_cycle(fgraph, ords):
-                raise InconsistencyError("Dependency graph contains cycles")
+                    for n in fgraph.apply_nodes:
```

Comment:
Add a comment that explain why we need this. Something like:

self.fail_validate can only be a hint that maybe/probably there is a cycle.
This is because inside replace() we could record many reason to not accept a change.
But we don't know which one will fail first inside validate().

Note, if you find in your benchmark that this is still taking times, we could speed it up again by making self.fail_validate a dict where values are the nodes that failed and the values the error. This would ask us to only revalidate the values of the dict and if they don't revalidate, we would return the error.